pep-9999.rst

```diff
@@ -0,0 +1,211 @@
+that are constructed with the ``covariant=True`` or ``contravariant=True``
+keyword arguments works well enough, it has downsides:
+
+- In PEP 484 it states:
+
+    Covariance or contravariance is not a property of a type variable,
+    but a property of a generic class defined using this variable. 
+
+  This is why type checkers do no allow using co- or contravariant type
+  variables outside of generic classes. 
+  So, instead of defining variance on the type variable itself, it
+  would make more sense to mark the type variable's variance when
+  defining the generic type, and only there. 
+
+- Because variance is a property of the generic type itself, type 
+  checkers do not allow using co- or contravariant type variables 
+  outside of generic classes. But as the variance is currently a
+  property of the type variables, they must be used within generic
+  classes.
+
+- Creating separate type variables for different variances often
+  results in up to 3x code duplication, especially for type variables
+  with constraints or an upper bound. 
+
+
+Proposal
+========
+
+
+All of these issues can be solved by overloading the ``+`` and ``-``
+prefix operators of ``typing.TypeVar`` instances, so that ``+T`` is an
+alias for a covariant type variable, and ``-T`` for a contravariant 
+one. These variance operators should only be used within the parameter
+list of ``typing.Generic`` and ``typing.Protocol`` when subclassing them.
+
+This syntax is practically identical to type parameter variance in the
+Scala programming language [1]_.
+
+
+
+Specification
+=============
+
+The new type variable variance syntax should be used within the generic
+parameter list of ``typing.Generic`` or ``typing.Protocol`` base classes,
+and only there. It is not allowed to be used on type variables that
+are already marked as covariant or contravariant, either with 
+``covariant=True``, ``contravariant=True`` or with the new ``+T`` or 
+``-T`` syntax.
+
+Simplified Syntax
+-----------------
+
+Instead of specifying the variance on the typevar itself::
+
+    T_co = typing.TypeVar('T_co', covariant=True)
+    T_contra = typing.TypeVar('T_contra', contravariant=True)
+
+    class SupportsInvertOld(typing.Protocol[T_co]):
+        def __invert__(self) -> T_co: ...
+
+    class SupportsPartialOrderOld(typing.Protocol[T_contra]):
+        def __le__(self, other: T_contra): ...
+
+
+The new syntax uses the ``+`` and ``-`` prefix operators to specify
+covariant and contravariant generic types inline::
+
+    T = typing.TypeVar('T')
+
+    class SupportsInvert(typing.Protocol[+T]):
+        def __invert__(self) -> T: ...
+
+    class SupportsPartialOrder(typing.Protocol[-T]):
+        def __le__(self, other: T): ...
+
+
+Differences with current syntax
+-------------------------------
+
+The new typevar operators return a transparent wrapper around the 
+original type variable, which can be accessed with the ``__orig__``
```

Comment:
Should this not be ``__origin__`` to match the other generic aliases for run time inspection?

Reply:
Ah, yes, good catch, that is what I was going for. I'll correct it.